src/main/java/underlay/Underlay.java

```diff
@@ -0,0 +1,168 @@
+package underlay;
+
+import blockchain.LightChainRMIInterface;
+import java.io.FileNotFoundException;
+import java.rmi.Naming;
+import java.rmi.RemoteException;
+import org.apache.log4j.Logger;
+import skipGraph.NodeInfo;
+import skipGraph.RMIInterface;
+import underlay.requests.*;
+import underlay.responses.*;
+import util.Util;
+
+
+
+/**
+ * The underlay class is used to abstract away the RMI primitives used in the skipGraph and
+ * lightchain nodes.
+ */
+public class Underlay {
+
+  /** The underlying RMIInterface instance. Is used for skipnode calls */
+  private RMIInterface targetRMI;
+
+  /** The underlying LightChainRMIInterface instance. Is used for lightchainnode class */
+  private LightChainRMIInterface targetLightChainRMI;
+
+  private final Logger logger = Logger.getLogger("");
+
+  public GenericResponse sendMessage(GenericRequest req, String targetAddress)
+      throws RemoteException, FileNotFoundException {
+    return sendMessage(req, targetAddress, InterfaceTypes.SkipNodeInterface);
+  }
+
+  /**
+   * This method returns a response after calling the underlying RMI implementation based on the
+   *  request type.
+   * @param req              The request
+   * @param targetAddress    A string which specifies the address of the target node
+   * @param type             The type (SkipNodeInterface or LightChainInterface)
+   * @return                 GenericResponse, the base class for all responses.
+   * @throws RemoteException
+   * @throws FileNotFoundException
+   */
+  public GenericResponse sendMessage(GenericRequest req, String targetAddress, InterfaceTypes type)
```

Comment:
I've changed the code based on your recommendations but I was wondering if the following would be even better:
Instead of the two methods you've outlined above calling the sendMessage function again, the sendMessage function with three parameters could instead call the methods above, ie the two parameter versions.

Reply:
Please implement as we discussed.

Reply:
https://github.com/yhassanzadeh13/lightchain-container/pull/15/commits/4de632c13888c1d13a2b27665e2f81c20988f596, in this commit we use the sendmessage function with two parameters all over the code.

Reply:
Changed/fixed in https://github.com/yhassanzadeh13/lightchain-container/pull/15/commits/e97c50a210f563c29f346cb17288ab76631f533c. It turns out just one sendmessage function with the genericrequest parameter is enough.